Allow cancellation of navigation events in Blazor

MackinnonBuck opened the pull request:
# Allow cancellation of navigation events in Blazor

Introduces new APIs that enable the cancellation of internal and external navigation events.

## API Overview

New `NavigationManager` APIs can be used to add and remove handlers that may asynchronously cancel internal navigations:

```csharp
public abstract class NavigationManager
{
    public void AddLocationChangingHandler(Func<LocationChangingContext, ValueTask> locationChangingHandler);
    public bool RemoveLocationChangingHandler(Func<LocationChangingContext, ValueTask> locationChangingHandler);
}

public class LocationChangingContext
{
    public string TargetLocation { get; }
    public bool IsNavigationIntercepted { get; }
    public CancellationToken CancellationToken { get; }
    public void PreventNavigation();
}
```

The `<NavigationLock />` component wraps these APIs so this functionality can easily be added to a Blazor component:

```razor
<NavigationLock OnBeforeInternalNavigation="OnBeforeInternalNavigation" ConfirmExternalNavigation="true|false" />

@code {
    private void OnBeforeInternalNavigation(LocationChangingContext context)
    {
        // ...
    }
}
```

When `ConfirmExternalNavigation` is `true`, the browser's built-in confirmation dialog prompts the user if they would like to continue before performing an external navigation.

## Important behavioral considerations

Since internal navigations can be canceled asynchronously, it's possible that multiple overlapping navigations may occur. These might include cases where the user is rapidly clicking the back button on a page or clicking multiple links before a decision about whether to permit the navigation is made. Following is a summary of how this asynchrony is handled:

* If there are any "location changing" handlers registered, all navigations are initially reverted, then replayed if the navigation does not get canceled.
* If overlapping navigation requests are made, the latest request always cancels earlier requests. This means:
  * The app may treat multiple back/forward button clicks as a single click.
  * If the user clicks multiple links before the navigation completes, the most recently clicked link will be navigated to.

It's also important to note that the `<NavigationLock />` component only takes effect after being rendered (just like any other Blazor component). Practically, this means it's possible that the user may quickly navigate past a page that was supposed to intercept the navigation, but it didn't because the user navigated away before the component had a chance to render.

Some other details to be aware of:
* Programmatically-initiated navigations (e.g. `NavigationManager.NavigateTo()`) will also invoke the "location changing" handlers.
* The navigation behavior for existing apps does not change if these features are not used.

Fixes #14962

MackinnonBuck committed:
Prototype for canceling internal navigations

MackinnonBuck committed:
New overlapping navigation behavior, changed APIs

MackinnonBuck committed:
Added APIs to block external navigations

MackinnonBuck committed:
Updated API baselines, added XML docs.

MackinnonBuck committed:
Update NavigationManager.cs

javiercn commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -262,6 +269,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// Adds a handler to process incoming navigation events.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to process incoming navigation events.</param>
+    public void AddLocationChangingHandler(IHandleLocationChanging locationChangingHandler)
+    {
+        AssertInitialized();
+        _locationChangingHandlers.Add(locationChangingHandler);
+        UpdateHasLocationChangingHandlers();
+    }
+
+    /// <summary>
+    /// Removes a previously-added location changing handler.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to remove.</param>
+    /// <returns><see langword="true"/> if the handler was removed, otherwise <see langword="false"/>.</returns>
+    public bool RemoveLocationChangingHandler(IHandleLocationChanging locationChangingHandler)
+    {
+        AssertInitialized();
+
+        if (_locationChangingHandlers.Remove(locationChangingHandler))
+        {
+            UpdateHasLocationChangingHandlers();
+            return true;
+        }
+
+        return false;
+    }
```
Is there a reason why this is not an `event`?

javiercn replied:
Is it because the callbacks are async?

MackinnonBuck replied:
@javiercn Yes, and also having an event it makes it easier to accidentally register an anonymous function as a `LocationChanging` handler and never unregister it.

javiercn replied:
@MackinnonBuck I can accept the first argument, but the second one is not really a valid concern. People is responsible for registering/unregistering events, the framework should not do anything about it.

I think we shouldn't add the interface, we should just take a `Func<LocationChangingContext, ValueTask>` in the methods. Otherwise, it makes it cumbersome to use the API.

MackinnonBuck replied:
@javiercn That sounds good to me. FWIW, there was a discussion on this topic previously: https://github.com/dotnet/aspnetcore/pull/24417#discussion_r512163217

MackinnonBuck committed:
Merge branch 'main' into mbuck/navigation-cancellation

MackinnonBuck committed:
Safeguard for overlapping navigations

Added a safeguard to ensure overlapping navigations are correctly canceled during high network latency scenarios.

SteveSandersonMS commented:
```cs
public CancellationToken CancellationToken { get; }
public bool IsCanceled { get; }
public void Cancel();
```

This triplet could be confusing, because we're using the word "cancel" to mean two different things (if I'm following correctly). The `CancellationToken` tells you if this `OnLocationChanging` call has been cancelled, whereas `Cancel`/`IsCanceled` are about the navigation itself being cancelled.

I'd suggest we disambiguate by not using the phrase "cancel" about the navigation itself, and saying something like "PreventNavigation" instead, e.g.:

```cs
public CancellationToken CancellationToken { get; }
public void PreventNavigation();
public bool IsNavigationPrevented { get; }
```

SteveSandersonMS commented:
Also to be honest, I'm not sure what is the use case for `IsNavigationPrevented`. It seems like the objective might be to do with orchestrating across multiple handlers, but given asynchrony, it doesn't quite cover it because even if you check once, another handler might then prevent navigation later.

What if instead, there was just `PreventNavigation` and `CancellationToken`, and the act of calling `PreventNavigation` would also cause the cancellation token to fire? In that case, other handlers would be notified that their opinion is no longer relevant since we're going to prevent navigation regardless, so they can stop doing work if they want to.

I'm not 100% sure how usable this would end up being (whether there are any cases where people would have to suppress a TaskCancelledException) but on the surface it seems like it might be a valid and simple pattern.

SteveSandersonMS commented:
More API thoughts:

>     public string Location { get; }

I'm guessing this is the destination location, not the current location, but just to be clear maybe rename to `DestinationUrl` or `TargetUrl` or similar (and I'd suggest "Url" over "location" since it matches the "Url" terminology on NavigationManager).

>    public bool ForceLoad { get; }

Do we need this info? If it's a client-originated navigation (`Blazor.navigateTo(...)`), I don't think the flag could ever be true, because if it was we'd have bypassed the navigation system anyway. So I think this can only be true for server-originated navigations, which might be a confusing subtlety. I'm not sure there are many cases where a developer needs this info to decide whether to allow the navigation to continue. If you have a concrete example then great, otherwise maybe we drop this flag (to avoid the subtlety mentioned above) and only add it later if a clear use case arises.

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+        {
+            for (var i = 0; i < handlerCount; i++)
+            {
+                await locationChangingHandlersCopy[i].OnLocationChanging(context);
```
This is interesting. We hadn't discussed whether the handlers would run in series or in parallel. (I don't mean on multiple threads BTW, I just mean with their async execution interleaved).

I had thought they would run in parallel but thinking about it now, I don't have any justification for why that should be.

Are there any pros/cons or important scenarios you're aware of that would push us to running them in series like this?

I guess one possible reason for running them in parallel would be if you had a handler that was something like this:

 1. Start saving the unsaved changes in the background
 2. If it takes too long or fails, bring up a dialog asking the user what they want to do

If you had more than one such handler, you might want them all to try doing their step 1 concurrently.

What do you think - are there any definite scenarios for requiring serial invocation?

MackinnonBuck replied:
I guess the only case I can think of for serial invocation is if you have multiple handlers intentionally registered in an order such that the more expensive handlers run last, after the simpler handlers have confirmed that they think the navigation should continue. But I would imagine you could just put all that logic in a single handler anyway, so I'm not sure if this is a real scenario.

I think running the handlers in parallel sounds good to me.

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <param name="forceLoad">Whether the navigation is attempting to bypass client-side routing.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation.</returns>
+    public async ValueTask<bool> NotifyLocationChanging(string uri, bool isNavigationIntercepted, bool forceLoad)
```
Does this have to be public? I suspect protected might suffice for the platform-specific subclasses to call it.

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// Sets whether there are active location changing handlers.
+    /// </summary>
+    /// <param name="value">Whether there are active location changing handlers.</param>
+    /// <returns><see langword="true"/> if the change was applied successfully, otherwise <see langword="false"/>.</returns>
```
Is the intention that platform-specific subclasses should override this to perform a call into JS to enable/disable the lock?

If so a name like `SetNavigationLockState(bool)` might be clearer. The fact that the lock state is determined based on having a nonzero number of handlers can be a private matter for NavigationManager.

Apologies if I'm misunderstanding the flow here!

SteveSandersonMS replied:
Also consider having the base class throw `NotSupportedException` here, so as to make it really clear that subclasses have to override it (but only if this isn't going to break people's existing unit tests or such like - will this get called implicitly even for existing code that doesn't use the navigation lock functionality?).

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// <summary>
+    /// Invokes <see cref="SetHasLocationChangingHandlers(bool)"/> if the existence of location changing handlers has changed.
+    /// </summary>
+    protected void UpdateHasLocationChangingHandlers()
```
Is this protected only because of the `AttachJsRuntime` flow for `RemoteNavigationManager`? If so is there any alternative? It seems quite overspecialised to make a protected method for this one thing if it's only relevant in one of the subclasses.

For example, if subclasses could directly read the "navigation locked" status, then `RemoteNavigationManager` could call its own `SetHasLocationChangingHandlers` from `AttachJsRuntime` if the navigation lock is meant to be active, and then `UpdateHasLocationChangingHandlers` could be private.

Or maybe that's not even needed. If `RemoteNavigationManager` receives the `SetHasLocationChangingHandlers` call too early, it could store the flag value and then use it during `AttachJsRuntime`.

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// Removes a previously-added location changing handler.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to remove.</param>
+    /// <returns><see langword="true"/> if the handler was removed, otherwise <see langword="false"/>.</returns>
```
What's the use case for this return value? What is a caller meant to do if we return false here?

MackinnonBuck replied:
This was mainly to follow the pattern that collections like `List<T>` use, where a `false` return value indicates that the element was not found in the collection and therefore wasn't removed. But for this specific feature, I'm not sure if there's much use in knowing whether the callback you've just removed was even added in the first place. Do you think it's fine to just make the return type `void`?

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// <param name="message">The message to display in the prompt. Note that some browsers will ignore this message for external navigations.</param>
+    /// <param name="externalNavigationsOnly">If <see langword="true"/>, the prompt will not display for internal page navigations.</param>
+    /// <returns>A <see cref="ValueTask"/> representing the completion of the operation.</returns>
+    protected internal virtual ValueTask EnableNavigationPromptAsync(string message, bool externalNavigationsOnly)
```
Are EnableNavigationPromptAsync/DisableNavigationPromptAsync really needed inside NavigationManager? I would have thought that the new `NavigationPrompt` component could register its own `beforeunload` handler via JS interop without involving the core NavigationManager.

Most likely I'm missing something, but it would be great to clarify. Thanks!

MackinnonBuck replied:
Since `NavigationPrompt` lives in `Microsoft.AspNetCore.Components.Routing`, using an `IJSRuntime` in the component would mean adding a reference to `Microsoft.JSInterop` in `Microsoft.AspNetCore.Components`. So, I added these APIs to `NavigationManager` to move the JS interop code out of `M.A.Components`.

An alternative could be moving `NavigationPrompt` to `Microsoft.AspNetCore.Components.Web`. What do you think?

SteveSandersonMS replied:
> An alternative could be moving NavigationPrompt to Microsoft.AspNetCore.Components.Web. What do you think?

Yep, that sounds great!

The high-level functionality here really is quite web-specific, particularly in the distinction between internal and external navigation (which we don't have any reason to think would have a corresponding equivalent in any particular non-web tech).

SteveSandersonMS commented on src/Components/Components/src/Routing/InternalNavigationLock.cs:
```diff
@@ -0,0 +1,57 @@
+/// <summary>
+/// A component that can be used to intercept internal navigation events.
+/// </summary>
+public class InternalNavigationLock : IComponent, IHandleLocationChanging, IDisposable
```
In terms of API obviousness, could we consider naming this `NavigationLock` and having it handle both the internal and external cases? Maybe we wouldn't need the `NavigationPrompt` component to be separate.

SteveSandersonMS commented on src/Components/Server/src/Circuits/RemoteNavigationManager.cs:
```diff
@@ -65,9 +67,14 @@ public void NotifyLocationChanged(string uri, string state, bool intercepted)
     /// <inheritdoc />
     [DynamicDependency(DynamicallyAccessedMemberTypes.PublicProperties, typeof(NavigationOptions))]
-    protected override void NavigateToCore(string uri, NavigationOptions options)
+    protected override async void NavigateToCore(string uri, NavigationOptions options)
```
`async void` is going to raise some eyebrows.

At the very least, this method ensure errors are handled (e.g., `try`/`catch` everything, and feed exceptions back into the system somehow).

Possibly better might be keeping `NavigateToCore` as `void`, but only calling it *after* we've been through some new optional async `BeforeNavigation` flow.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -7,39 +7,82 @@ import { EventDelegator } from '../Rendering/Events/EventDelegator';
+
+function onBeforeUnload(event: BeforeUnloadEvent) {
+  event.preventDefault();
+  return event.returnValue = navigationPromptOptions?.message;
```
Is this intended as shorthand for assignment and then return of the same value? If so would it be possible to split this into two statements for clarity?

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -116,20 +178,111 @@ function performInternalNavigation(absoluteInternalHref: string, interceptedLink
+    currentHistoryIndex++;
+    history.pushState({
+      userState: state,
+      index: currentHistoryIndex,
```
Very minor nit-pick: consider naming this `_index` to just give us a justification for expecting people to know this is an internal implementation detail we might change in the future!

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -116,20 +178,111 @@ function performInternalNavigation(absoluteInternalHref: string, interceptedLink
+    await navigateHistoryWithoutPopStateCallback(delta);
+  } else {
+    // Allow the user to cancel the navigation, if necessary.
+    const navigationCanceled = displayNavigationPromptIfEnabled();
```
I'm still wondering if this has to be baked into NavigationManager itself. Would it be practical for the NavigationLock component to deal with the `beforeunload` event handler stuff completely separately? That is, everything to do with external navigations.

Or is there some necessity for NavigationManager to have a concept of external navigations?

MackinnonBuck replied:
I've pulled out the `<NavigationLock />`-specific code into its own `.ts` file.

MackinnonBuck committed:
PR feedback

MackinnonBuck committed:
Cleanup

MackinnonBuck committed:
Create blazor.webview.js

MackinnonBuck committed:
Fixed deadlock, better exception handling

MackinnonBuck committed:
Minor cleanup

MackinnonBuck commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs:
```diff
@@ -148,6 +149,8 @@ internal async Task RunAsyncCore(CancellationToken cancellationToken, WebAssembl
             var jsComponentInterop = new JSComponentInterop(_rootComponents.JSComponents);
             _renderer = new WebAssemblyRenderer(Services, loggerFactory, jsComponentInterop);
 
+            WebAssemblyNavigationManager.Instance.CreateLogger(loggerFactory);
```
I'm not a fan of having to do it this way, but it seemed like one of the only options given the `WebAssemblyNavigationManager` instance is a static singleton unassociated with DI. I'm definitely open to alternative mechanisms for logging exceptions thrown in "location changing" events.

javiercn replied:
I don't remember why it is done this way, maybe we could change it to be created from DI and that way it only needs to get injected in the constructor.

halter73 commented:
@MackinnonBuck @SteveSandersonMS Do we want to have this go through an API review?

MackinnonBuck commented:
@halter73 Yes, we'll want to do an API review for this.

MackinnonBuck committed:
BlazorWebView support

javiercn commented on src/Components/Server/src/Circuits/RemoteNavigationManager.cs:
```diff
@@ -65,6 +72,31 @@ public void NotifyLocationChanged(string uri, string state, bool intercepted)
         NotifyLocationChanged(intercepted);
     }
 
+    public void HandleLocationChanging(int callId, string uri, bool intercepted)
+    {
+        NotifyLocationChanging(uri, intercepted, result =>
+        {
+            bool success;
+
+            if (result.Exception is { } exception)
+            {
+                Log.NavigationFailed(_logger, uri, exception);
+                success = false;
+            }
+            else if (result.Canceled)
+            {
+                Log.NavigationCanceled(_logger, uri);
+                success = false;
+            }
+            else
+            {
+                success = true;
+            }
+
+            _jsRuntime.InvokeVoidAsync(Interop.EndLocationChanging, callId, success).Preserve();
+        });
+    }
```
Why is this done this way vs how we do location changed in the server (see https://github.com/dotnet/aspnetcore/blob/main/src/Components/Server/src/ComponentHub.cs#L288-L297)

I believe this has implications with regards to things like how exceptions are dealt with when they happen at different stages.

javiercn commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +282,87 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <param name="continueNavigation">An action that continues the navigation based on the provided <see cref="LocationChangingResult"/>.</param>
+    protected void NotifyLocationChanging(string uri, bool isNavigationIntercepted, Action<LocationChangingResult> continueNavigation)
```
This signature doesn't look right, how would it look if this was async all the way through? (See OnLocationChanged for reference in ComponentHub)

MackinnonBuck replied:
The previous implementation was completely async, but the problem is that on Blazor Server, the SignalR property `MaximumParallelInvocationsPerClient` is set to 1. This effectively makes things like JS interop calls blocked until this method completes (and if you try to do a JS interop call in an "OnLocationChanging" handler, the app will deadlock). This is the reason for making this completely synchronous and having a separate "end location changing" call to let the JS side know about the result.

javiercn replied:
Would it work if it doesn't do this via JS interop but via methods in the hub?

SteveSandersonMS replied:
I agree it seems like it should be possible to hide the SignalR-serial-calls limitation from `NavigationManager` subclasses. If this method was `async`, then perhaps the hub method could deal with awaiting that task and then calling the completion (but without actually returning the child task back to SignalR, so it doesn't block).

javiercn commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +282,87 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+            locationChangingTasks[i] = locationChangingHandlersCopy[i].Invoke(context);
+        }
+
+        Task.WhenAll(locationChangingTasks.Take(handlerCount)).WaitAsync(cancellationToken).ContinueWith(t =>
+        {
+            continueNavigation(new(cancellationToken.IsCancellationRequested, t.Exception));
+        }, TaskScheduler.Current);
```
@MackinnonBuck this can be problematic. `ContinueWith` != `await task`, for example the execution context is not flowed via `ContinueWith`. 

I would suggest trying to always rely on async/await when possible and only skip doing that if you have a really good reason to do so.

In this case, you could move the code to a local helper method and just call it from here.

SteveSandersonMS replied:
> The key is that we don't await the task on the hub, we will just "deliver" the locationchanging event.

How do we avoid awaiting? If there are multiple handlers, something has to `WhenAll` them, doesn't it?

javiercn commented on src/Components/Components/src/PublicAPI.Unshipped.txt:
```diff
@@ -1,10 +1,21 @@
+Microsoft.AspNetCore.Components.Routing.LocationChangingContext.IsNavigationIntercepted.get -> bool
+Microsoft.AspNetCore.Components.Routing.LocationChangingContext.PreventNavigation() -> void
+Microsoft.AspNetCore.Components.Routing.LocationChangingContext.TargetLocation.get -> string!
+Microsoft.AspNetCore.Components.Routing.LocationChangingResult
+Microsoft.AspNetCore.Components.Routing.LocationChangingResult.Canceled.get -> bool
+Microsoft.AspNetCore.Components.Routing.LocationChangingResult.Exception.get -> System.AggregateException?
```
Based on what I read so far, we might be able to get rid of this by going async all the way through on NavigationManager.OnLocationChanging and just returning `bool` from the method to signal the navigation was cancelled.

MackinnonBuck replied:
> we might be able to get rid of this by going async all the way through on NavigationManager.OnLocationChanging

This is the part I'm still unsure about, per my reasoning in [this comment](https://github.com/dotnet/aspnetcore/pull/42638#discussion_r920372967).

javiercn commented on src/Components/Components/src/Routing/LocationChangingContext.cs:
```diff
@@ -0,0 +1,43 @@
+{
+    private readonly CancellationTokenSource _cts;
+
+    internal LocationChangingContext(string destinationUrl, bool isNavigationIntercepted, CancellationTokenSource cts)
```
Does this also need to receive the state?

MackinnonBuck replied:
Yeah, that seems reasonable. I wonder if we should add this in a follow-up just to keep the initial API/implementation as simple as possible?

javiercn replied:
I think you need to handle both at the same time, adding a feature can't break another feature.

MackinnonBuck replied:
How would this be breaking another feature? The state will still be provided in `LocationChangedEventArgs`. I interpreted your suggestion as meaning we would also add a `public string? HistoryEntryState { get; }` property to `LocationChangingContext` so that you could see the state of the location being navigated to. Couldn't this be something we add in a follow-up without breaking other things? Let me know if I'm misinterpreting all of this 😄

javiercn commented on src/Components/Server/src/Circuits/RemoteNavigationManager.cs:
```diff
@@ -65,6 +72,31 @@ public void NotifyLocationChanged(string uri, string state, bool intercepted)
         NotifyLocationChanged(intercepted);
     }
 
+    public void HandleLocationChanging(int callId, string uri, bool intercepted)
```
This should probably be async and have it await it on the CircuitHost calling code

MackinnonBuck replied:
I believe this ties back into [this comment](https://github.com/dotnet/aspnetcore/pull/42638#discussion_r920372967), but let me know if I'm mistaken.

MackinnonBuck replied:
@javiercn Ah, ok yes, I get it now. I've just updated everything - please let me know if it differs from what you had in mind.

/cc @SteveSandersonMS

javiercn commented on src/Components/WebAssembly/WebAssembly/src/Infrastructure/JSInteropMethods.cs:
```diff
@@ -29,4 +29,13 @@ public static void NotifyLocationChanged(string uri, string? state, bool isInter
+    /// <summary>
+    /// For framework use only.
+    /// </summary>
+    [JSInvokable(nameof(NotifyLocationChanging))]
+    public static void NotifyLocationChanging(int callId, string uri, bool isInterceptedLink)
+    {
+        WebAssemblyNavigationManager.Instance.HandleLocationChanging(callId, uri, isInterceptedLink);
+    }
```
Ah, I think this is the reason. These APIs target the static fields/values.

Maybe we need to file an issue here to see if we could convert some of these things to use JSObjectReferences for dispatching

MackinnonBuck replied:
This is something I thought about while implementing this feature - I wonder we could simplify things if we had a `DotNetObjectReference` to some object responsible for invoking the "location changing" handlers. Then we wouldn't have to separate out things into "notify location changing" and "end location changing" steps since a similar mechanism is handled behind the scenes in the `JSRuntime`. But maybe there's some reason why this wouldn't work. Is this approach what you had in mind?

javiercn replied:
@MackinnonBuck something like that might work. I would suggest you file an issue and give it a try in the future if you want if you have spare cycles. (After RC1)

MackinnonBuck commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +282,87 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+        var context = new LocationChangingContext(uri, isNavigationIntercepted, _locationChangingCts);
+        var cancellationToken = _locationChangingCts.Token;
+        var handlerCount = _locationChangingHandlers.Count;
+        var locationChangingHandlersCopy = ArrayPool<LocationChangingHandler>.Shared.Rent(handlerCount);
```
I think having only one handler is a pretty common case. We could certainly add a hot path.

MackinnonBuck commented on src/Components/WebView/WebView/src/Services/WebViewNavigationManager.cs:
```diff
@@ -20,8 +28,56 @@ public void LocationUpdated(string newUrl, string? state, bool intercepted)
+    }
+
     protected override void NavigateToCore(string uri, NavigationOptions options)
     {
-        _ipcSender.Navigate(uri, options);
+        NotifyLocationChanging(uri, false, result =>
+        {
+            if (result.Exception is { } exception)
+            {
+                Log.NavigationFailed(_logger, uri, exception);
+            }
+            else if (result.Canceled)
+            {
+                Log.NavigationCanceled(_logger, uri);
+            }
+            else
+            {
+                _ipcSender.Navigate(uri, options);
+            }
+        });
+    }
```
Possibly, but they couldn't know right away since the navigation isn't guaranteed to complete by the time `NavigateTo` returns. Were you thinking we could add some other event to `NavigationManager` like `NavigationCanceled`? If so, I think that should be something we add as a follow-up to these changes.

javiercn replied:
I was thinking we could add `ValueTask<bool> NavigateToAsync` which lets you know if the navigation is cancelled (or succeeds in this case), and NavigateTo which you use if you just don't care. 

If you navigate programmatically but that does nothing, I think you want to be able to react in code. Specially since the navigation might take some time to take effect.

SteveSandersonMS replied:
Follow up:

@javiercn and I were discussing this and trying to decide if it's important to have in 7.0 or whether it would be OK to leave until we have a more concrete requirement from developer scenarios. We didn't get a definite conclusion.

I've been trying to think of the most plausible scenario for some code genuinely needing to know whether the navigation was cancelled or not, and it's quite hard to think of one. The closest I've got is some flow where an action button becomes disabled once you press it, like this:

```cs
private async Task SaveChangesAsync() // Invoked via a button click
{
    this.disableSaveButton = true;
    if (await myData.SaveAsync())
    {
        // Save succeeded; redirect away
       NavManager.NavigateTo("list");
    }
    else
    {
        // Save failed - re-enable the button
        this.disableSaveButton = false;
    }
}
```

If the `NavigateTo` call gets cancelled, the user will be stuck because the button isn't re-enabled.

However I don't think this is a 100% valid scenario as presented here, because

1. There's still the question of *why would some completely unrelated code be interfering with navigation*? Normally it only makes sense for the code that manages the edit flow to block the navigation.
2. Even if you did re-enable the save button, that's probably not what you want to do (they have already saved the changes by now). You probably want to retry the navigation or something, but it's very unclear what sort of user gesture would trigger that. TBH the entire UX flow doesn't really make sense if the navigation might randomly not happen.

So I'm still on the lookout for a convincing account of a realistic UI that would make use of the `NavigateToAsync` result value. Maybe there's a simple example someone can come up with!

MackinnonBuck committed:
Started E2E tests, fixed Blazor.navigateTo() bug

Blazor.navigateTo() wasn't invoking the location changing callback. This is now fixed.

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -1,10 +1,14 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers;
+using System.Linq;
```
Is `System.Linq` needed? Pranav went on a whole campaign of removing these references from low-level code in the past.

SteveSandersonMS replied:
I guess it's for the `locationChangingTasks.Take` call. Might be possible to use an array segment or Span instead.

MackinnonBuck replied:
Yeah, this was something I wasn't too happy about either. Unfortunately, `Task.WhenAll` only has overloads accepting `Task[]` and `IEnumerable<Task>`. So even if we use an `ArraySegment<Task>`, the extra allocation to get an `IEnumerable<Task>` seems inevitable. We could also consider not using `ArrayPool<Task>` at all and just allocating a `Task[]` with a hot path for the single handler scenario.

MackinnonBuck replied:
> We could also consider not using `ArrayPool<Task>` at all and just allocating a `Task[]` with a hot path for the single handler scenario.

I've gone ahead and changed it to this - let me know if you disagree with that choice!

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +282,87 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+        }, TaskScheduler.Current);
+
+        ArrayPool<LocationChangingHandler>.Shared.Return(locationChangingHandlersCopy);
+        ArrayPool<Task>.Shared.Return(locationChangingTasks);
```
These should be in a `finally`

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +282,87 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// </summary>
+    /// <param name="value">Whether there are active location changing handlers.</param>
+    protected virtual void SetNavigationLockState(bool value)
+        => throw new NotSupportedException();
```
```suggestion
        => throw new NotSupportedException($"To support navigation locks, {nameof(NavigationManager)} subclasses must override {nameof(SetNavigationLockState)");
```

SteveSandersonMS commented on src/Components/Web/src/Routing/NavigationLock.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        var confirmExternalNavigation = ConfirmExternalNavigation;
+        if (confirmExternalNavigation != lastConfirmExteranlNavigation)
```
```suggestion
        if (confirmExternalNavigation != lastConfirmExternalNavigation)
```

SteveSandersonMS commented on src/Components/Web.JS/src/NavigationLock.ts:
```diff
@@ -0,0 +1,37 @@
+    window.addEventListener('beforeunload', onBeforeUnload);
+  }
+
+  numLocksWithPromptEnabled++;
```
Not essential, but instead of reference counting, it might be simpler or more robust to have each NavigationLock instance make up a GUID for itself (or some other ID) and send that to the JS side which can track all of the ones with an active lock in a `Map`. Then we don't have to be concerned about double-registration or double-unregistration.

MackinnonBuck commented on src/Components/Web/src/Routing/NavigationLock.cs:
```diff
@@ -0,0 +1,89 @@
+    {
+    }
+
+    async Task IComponent.SetParametersAsync(ParameterView parameters)
```
The only reason I did this was to avoid adding these methods to the public API. Do you think that's a valid concern or is it fine to change it?

SteveSandersonMS replied:
In most cases for equivalent components we did regular interface implementations (e.g., `RouteView`). However I don't know of any reason why it would be a problem to do it like you have. Either way is fine with me.

javiercn replied:
I would follow the established pattern unless there is an explicit concrete benefit of using an alternative.

MackinnonBuck committed:
PR feedback

MackinnonBuck committed:
Miscellaneous improvements

* Added hot path for case where there is only a single location changing handler
* Simplified navigation manager logic so it no longer needs to be aware of SignalR invocation limits
* Hopefully improved test stability

MackinnonBuck commented on src/Components/Server/src/Circuits/CircuitHost.cs:
```diff
@@ -536,6 +536,31 @@ public async Task OnLocationChangedAsync(string uri, string state, bool intercep
+        }
+        catch (Exception ex)
+        {
+            // Exceptions thrown by location changing handlers should be treated like unhandled exceptions in user-code.
```
I wonder if we should really be treating exceptions thrown in location changing handlers as non-fatal? In the current WebAssembly implementation of this feature, the navigation is canceled if an exception is thrown in the location changing handler. Should we treat it as a critical failure there instead? And, if we decide to make this case non-fatal, should we let the navigation continue if no other handlers cancel it?

SteveSandersonMS replied:
I'd err on the side of treating unhandled exceptions as fatal unless there's a definite use case in which developers can't be expected to avoid unhandled exceptions. In this case I expect a developer could use try/catch inside their handler if they are doing something that might fail.

Generally in WebAssembly we've had a more laissez-faire attitude because we're inheriting the JS semantics around errors being something you log but may be able to continue afterwards. On Server, it's much more important to have clearly defined rules about when you can and can't continue after an exception.

> Should we treat it as a critical failure there instead?

If possible, that would be nice. But as long as something ends up in the JS console giving the exception details, that's enough. It just has to be reasonable for the developer to see what went wrong.

MackinnonBuck committed:
Switch NavigationLock to use IDs

MackinnonBuck committed:
Update RoutingTest.cs

MackinnonBuck committed:
Fixed history navigation test

MackinnonBuck committed:
Update NavigationManagerComponent.razor

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,111 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+        {
+            if (handlerCount == 1)
+            {
+                await _locationChangingHandlers[0].Invoke(context);
```
Looks like the `handlerCount == 1` case has slightly different cancellation behavior than then `handlerCount != 1` case.

On this branch, we're not short-circuiting the `await` if the cancellation token fires, whereas on the other branch we do.

I don't fully know what difference it will make to the upstream caller. But should we bring these two branches into alignment?

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,111 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+
+            return !cancellationToken.IsCancellationRequested;
+        }
+        catch (TaskCanceledException ex) when (ex.CancellationToken == cancellationToken)
```
In other cases we've made a point of avoiding exception filters because they had poor perf/behavior characteristics on WebAssembly AOT. @lewing Do you know if that is fixed these days - are we free to use exception filters now, or is it still better to avoid them?

Example: See `ComponentBase` inside `RunInitAndSetParametersAsync`.

If it's just as easy to do the equality check inside the `catch` and manually rethrow, that might be better.

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,111 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+    /// Removes a previously-added location changing handler.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to remove.</param>
+    /// <returns><see langword="true"/> if the handler was removed, otherwise <see langword="false"/>.</returns>
```
```suggestion
```

SteveSandersonMS commented on src/Components/Components/src/Routing/LocationChangingContext.cs:
```diff
@@ -0,0 +1,43 @@
+    public bool IsNavigationIntercepted { get; }
+
+    /// <summary>
+    /// Gets a <see cref="System.Threading.CancellationToken"/> that can be used to determine if this navigation was canceled.
```
```suggestion
    /// Gets a <see cref="System.Threading.CancellationToken"/> that can be used to determine if this navigation was canceled (for example,
    /// because the user has since triggered a different navigation).
```

SteveSandersonMS commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -1,10 +1,13 @@
 
 namespace Microsoft.AspNetCore.Components;
 
+using LocationChangingHandler = Func<LocationChangingContext, Task>;
```
NBD either way.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -116,18 +153,97 @@ function performInternalNavigation(absoluteInternalHref: string, interceptedLink
+      resolve();
+    };
+
+    history.go(delta);
```
Could you clarify why we do this thing of using the event handler itself to temporarily remove and put back the event handler?

A possibly-simpler alternative would be just setting a flag saying "ignore popstate callbacks" before we call `history.go` and then clearing the flag after. Or is there some case where one navigation synchronously triggers another or something, causing us to need multiple levels of flag?

MackinnonBuck replied:
My idea behind doing it this way was to avoid situations where some part of the code assumes that the navigation has completed and the `location` and `history` objects are up-to-date (because `history.go()` is asynchronous). For example, we probably want to wait until the navigation completes before calling `notifyLocationChanged`: https://github.com/dotnet/aspnetcore/blob/33653224a3000cfa94dde308919024228262f652/src/Components/Web.JS/src/Services/NavigationManager.ts#L222-L239
There are of course other ways to achieve this, but I thought making a helper enabling `async`/`await` would be easier to read and work with than having additional global state/callbacks keeping track of where we are in the navigation. However, I'm definitely open to considering alternative approaches that reduce the complexity of all this!

SteveSandersonMS reviewed (approved):
Amazing job, @MackinnonBuck!

I've added some comments but none of them should significantly impact the implementation, so I'm fine with this being merged whenever you're ready (especially if you're able to resolve most of those comments easily).

The one part I haven't fully internalised is the `NavigationManager.ts` changes since there are quite a lot of overlapping flows, so it's quite hard to be sure I'm interpreting everything that could possibly happen. Given the good test coverage and the remaining RC releases I'm OK with that for now.

MackinnonBuck committed:
PR feedback

egil commented on src/Components/Components/src/Routing/LocationChangingContext.cs:
```diff
@@ -0,0 +1,43 @@
+{
+    private readonly CancellationTokenSource _cts;
+
+    internal LocationChangingContext(string destinationUrl, bool isNavigationIntercepted, CancellationTokenSource cts)
```
Thanks for the ping Steve. 

I don't think its a problem in this case, as it looks as if `NavigationManager` does all the heavy lifting. bUnits [`FakeNavigationManager`](https://github.com/bUnit-dev/bUnit/blob/main/src/bunit.web/TestDoubles/NavigationManager/FakeNavigationManager.cs) inherits from `NavigationManager` and overrides ` NavigateToCore`. In ` NavigateToCore` we set the `Uri` property and call `NotifyLocationChanged`. If any of the navigation manager implementations in Blazor does not new-up `LocationChangingContext`, then there is no problem as far as I can tell. It looks like we would just need to change our code to call `NotifyLocationChangingAsync` instead.

That said, I'm checking this from my phone while on vacation, so cannot be 100% sure. What do you think @linkdotnet?

linkdotnet replied:
From the usages of `LocationChangingContext` inside `NavigationManager` it does not look like we need it public as you said @egil.

MackinnonBuck committed:
Debugging test failures in CI

MackinnonBuck committed:
Updated E2E tests

Hopefully these changes make them pass in CI

MackinnonBuck committed:
Fix potential trimming issue

MackinnonBuck committed:
History state entry support, more tests

javiercn commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,130 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+
+                    for (var i = 0; i < handlerCount; i++)
+                    {
+                        var handlerTask = locationChangingHandlersCopy[i].Invoke(context);
```
Same comment as above, I believe you can drop the Invoke

javiercn commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,130 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+                            return false;
+                        }
+
+                        locationChangingTasks[i] = handlerTask.AsTask();
```
I don't think using `ValueTask` here adds a lot of value over using `Task` directly. There is no return value so a handler can do return Task.CompletedTask if their handler executes synchronously and we are having to do `AsTask()` in all cases but the simplest one.

javiercn commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +281,130 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+
+        _locationChangingCts = new();
+
+        var context = new LocationChangingContext(uri, state, isNavigationIntercepted, _locationChangingCts);
```
Is the cancellation token being used by the callbacks to prevent the navigation? Are there any other cases you can think of where we follow this pattern?

javiercn commented on src/Components/Web/src/Routing/NavigationLock.cs:
```diff
@@ -0,0 +1,85 @@
+        await OnBeforeInternalNavigation.InvokeAsync(context);
+    }
+
+    async ValueTask IAsyncDisposable.DisposeAsync()
+    {
+        NavigationManager.RemoveLocationChangingHandler(OnLocationChanging);
+        await JSRuntime.InvokeVoidAsync(NavigationLockInterop.DisableNavigationPrompt, _id);
+    }
```
```suggestion
    ValueTask IAsyncDisposable.DisposeAsync()
    {
        NavigationManager.RemoveLocationChangingHandler(OnLocationChanging);
        return JSRuntime.InvokeVoidAsync(NavigationLockInterop.DisableNavigationPrompt, _id);
    }
```

MackinnonBuck replied:
Is this better for performance reasons? I've formed a habit of avoiding directly returning tasks for the reasons [outlined here](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task) (although it looks like some of those reasons don't apply to this case).

javiercn commented on src/Components/test/testassets/BasicTestApp/RouterTest/ConfigurableNavigationLock.razor:
```diff
@@ -0,0 +1,92 @@
+@{
+    var onBeforeInternalNavigation = _blockInternalNavigation
+        ? BlockNavigationAsync
+        : default(Func<LocationChangingContext, Task>);
```
You can probably put everything in a single line too

javiercn replied:
ah, I see. leaving it like this is ok then.

javiercn reviewed (commented):
Looks like great progress!

I have made several comments on the code and I will give a bit more high level feedback here:
* I would expect to see some more tests, specifically unit tests on the NavigationManager behavior.
* I would expect to see tests that validate the behavior when a handler throws or behaves in a bad way.
  * For example, it will be good to have tests that validate the behavior when there is a long running handler and we trigger multiple navigations from the client, and similar tests.

In terms of the callback design, I am not sure how to feel about using the cancellation token source to signal the navigation being blocked for two reasons:
* I think it is easier to spot that you need to return a value than dig through the context to find and call `Prevent()`
* It feels like the cancellation token source is being `abused` to signal a short-circuit in the navigation. It relies on exceptions for control flow, when all handlers have a way deal with a cancellation without throwing an exception (if they choose to do so).
* The CancellationToken should probably be a parameter on the callback instead of a property in the context (it is most common to find it like that in .NET APIs)

It is clever code, and it takes a while to understand why it works as the relationship between the cancellation token and the result is implicit.

I would much rather see the callback as `Func<LocationChangingContext, CancellationToken, Task<bool>>` as it more clearly reflects the contract in the signature.

MackinnonBuck committed:
Some PR feedback

MackinnonBuck committed:
PR feedback + added NavigationManager unit tests

MackinnonBuck commented:
@javiercn I've addressed some of your feedback around cancellation and added some unit tests (with a few more still remaining).

/cc @SteveSandersonMS in case you have any thoughts about the recent modifications.

javiercn commented on src/Components/Components/test/NavigationManagerTest.cs:
```diff
@@ -208,6 +215,356 @@ public void GetUriWithQueryParameters_ThrowsWhenAnyParameterNameIsEmpty(string b
+
+        static async ValueTask HandleLocationChanging(LocationChangingContext context)
+        {
+            await Task.Delay(1000);
```
This will introduce flakyness in the build.

You can use a TaskCompletionSource instead to deterministically decide when to complete the handler task. (See other tests in Renderer if you want examples)

javiercn replied:
I think what you want to do here is similar to what we do in the Renderer tests. Use a TCS on the handler, call
`var navigation1 = navigationManager.RunNotifyfLocationChangingAsync($"{baseUri}/subdir1", null, false);`

Assert that it hasn't completed (as you would expect). Set the result on the TCS and assert that it completed and produced the result you expected. 

You shouldn't need to have the Timeout

javiercn commented on src/Components/Components/test/NavigationManagerTest.cs:
```diff
@@ -224,9 +581,16 @@ public new void Initialize(string baseUri, string uri)
             base.Initialize(baseUri, uri);
         }
 
+        public async Task<bool> RunNotifyfLocationChangingAsync(string uri, string state, bool isNavigationIntercepted)
```
typo

javiercn commented on src/Components/Components/test/NavigationManagerTest.cs:
```diff
@@ -208,6 +215,356 @@ public void GetUriWithQueryParameters_ThrowsWhenAnyParameterNameIsEmpty(string b
+        {
+            try
+            {
+                await Task.Delay(System.Threading.Timeout.Infinite, context.CancellationToken);
```
You can do this with a TaskCompletionSource too.

```
var tcs = new TaskCompletionSource();
ct.Register(() => tcs.SetCanceled());
```

and inside this method you do await tcs.Task;

MackinnonBuck replied:
I think the problem is that if the `CancellationToken` gets canceled before the location changing callback executes, then the TCS will never get canceled (and this actually happens in the test because the overlapping navigations are fired off synchronously). Do you think it's fine to leave as-is?

javiercn replied:
Doesn't it work if you do `context.CancellationToken.ThrowIfCancelled()` before registering the cancellation?

javiercn replied:
That said, I am somewhat ok leaving it as it is.

javiercn commented on src/Components/Components/test/NavigationManagerTest.cs:
```diff
@@ -208,6 +215,356 @@ public void GetUriWithQueryParameters_ThrowsWhenAnyParameterNameIsEmpty(string b
+
+        static async ValueTask HandleLocationChanging_PreventNavigation(LocationChangingContext context)
+        {
+            await Task.Delay(1000);
```
```suggestion
            await Task.Yield();
```

javiercn commented on src/Components/Components/test/NavigationManagerTest.cs:
```diff
@@ -1,10 +1,17 @@
+
 namespace Microsoft.AspNetCore.Components;
 
 public class NavigationManagerTest
```
Tests look great and thorough; I would suggest adding a couple of additional tests where the handler throws an exception both synchronously and asynchronously to capture that behavior, since it is something that will likely happen on users apps.

MackinnonBuck replied:
Just added some tests that do this

javiercn commented on src/Components/Components/src/Routing/LocationChangingContext.cs:
```diff
@@ -38,13 +37,13 @@ internal LocationChangingContext(string targetLocation, string? historyEntryStat
     /// (for example, because the user has triggered a different navigation).
     /// </summary>
-    public CancellationToken CancellationToken => _cts.Token;
+    public CancellationToken CancellationToken { get; }
```
I think we should consider passing the cancellation token as a separate argument to the handler, there is no relationship between the cancellation token and the rest of the data in this class.

In addition to that, it is "interesting" that this class looks "identical" to `LocationChangedEventArgs` except for the CancellationToken in this one (and the method to prevent the navigation). 

I am not suggesting that we "reuse" the same class, but an indicator that the cancellation token should probably not be here.

MackinnonBuck replied:
@javiercn The challenge with having the `CancellationToken` as its own argument in the callback is that `EventCallback<TValue>` only allows there to be a single parameter, so if you define a location changing handler using `<NavigationLock OnBeforeInternalNavigation="Foo" />`, then `Foo()` can only accept a single argument (the `LocationChangingContext`). So the only option in that case is to leave the `CancellationToken` as a property on `LocationChangingContext`.

I guess one way we could make this work is that we have a separate `LocationChangingEventArgs` whose two properties are the `LocationChangingContext` and the `CancellationToken`. So, if you add a location changing handler using the `NavigationManager` API, your callback will still have a separate argument for the `CancellationToken`. I'm not sure how I feel about having to do `args.Context.PreventNavigation()` rather than just `context.PreventNavigation()` though.

Thoughts?

MackinnonBuck replied:
And also, the limitations of `EventCallback<TValue>` are also a reason why having a `PreventNavigation()` method could be preferred over relying on a return value from the handler (since event callbacks cannot return anything other than a `Task`).

javiercn commented on src/Components/Components/src/Routing/LocationChangingContext.cs:
```diff
@@ -8,17 +8,16 @@ namespace Microsoft.AspNetCore.Components.Routing;
-    private readonly CancellationTokenSource _cts;
-
-    internal LocationChangingContext(string targetLocation, string? historyEntryState, bool isNavigationIntercepted, CancellationTokenSource cts)
+    internal LocationChangingContext(string targetLocation, string? historyEntryState, bool isNavigationIntercepted, CancellationToken cancellationToken)
```
One thing to consider here, is how would people write unit tests for this. Does this type need to be "constructible" by people in unit tests?

MackinnonBuck replied:
I believe this discussion was had [here](https://github.com/dotnet/aspnetcore/pull/42638#discussion_r923501645). The conclusion seems to be that it's fine if this stays `internal`, at least for the bUnit case.

javiercn commented on src/Components/Server/src/Circuits/RemoteNavigationManager.cs:
```diff
@@ -89,21 +89,28 @@ protected override void NavigateToCore(string uri, NavigationOptions options)
+            catch (Exception ex)
             {
-                _jsRuntime.InvokeVoidAsync(Interop.NavigateTo, uri, options).Preserve();
+                Log.NavigationFailed(_logger, uri, ex);
```
What happens when a handler throws in this scenario compared to when the navigation is started from JS?

MackinnonBuck replied:
At the time you wrote this comment, it would log an error but not tear down the circuit. I've since changed that behavior so that it tears down the circuit for programmatically initiated navigations like it does for JS-initiated navigations.

javiercn reviewed (commented):
Changes look great, I have a couple of open questions/suggestions left, but this will be ready to go after resolving those.

MackinnonBuck committed:
Updated E2E tests

javiercn commented:
> UPDATE: Actually, the unfortunate thing about having more than two arguments in the location changing handler is that EventCallback<TValue> only allows for a single parameter. So we would have to re-wrap the LocationChangingContext and CancellationToken into a single object before so it can be passed to the OnBeforeInternalNavigation event callback when using the <NavigationLock/> component. And since using <NavigationLock/> is arguably the primary way people are going to implement location changing handlers, I think it's probably best to leave CancellationToken as a property on LocationChangingContext

I think I get what you are saying and that's a valid reason to keep the cancellation token in the same place.

MackinnonBuck committed:
Improved exception handling, more unit tests

MackinnonBuck committed:
Update TestCircuitHost.cs

javiercn commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -274,6 +279,159 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+
+                        if (completedHandler.Exception is { } exception)
+                        {
+                            throw exception;
```
```suggestion
                            await completedHandler;
```

`throw task.exception` can have different semantics (I believe it overrides the StackTrace and maybe other properties). That is why in some places of the codebase we use [`ExceptionDispatchInfo.Capture(exception).Throw()`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?view=net-6.0#remarks) to preserve the original information.

If you simply await the task, the async machinery takes care of making sure that everything gets properly propagated.

javiercn reviewed (approved):
This looks great!

Thanks @MackinnonBuck for persevering through all the nuances on the implementation details.

We still need to solve the problem we are discussing offline, but I am signing off already as I think we will be good to go after that.

MackinnonBuck committed:
Improved exception handling in handler callbacks

MackinnonBuck committed:
Fixed/verified Blazor Hybrid functionality

SteveSandersonMS commented:
> /cc @SteveSandersonMS in case you have any thoughts about the recent modifications.

It looks great to me. Reading through the comments, it seems like this has been worked out in great detail now. Thanks also to @javiercn for the quality feedback.

MackinnonBuck commented:
I'm going to merge this now - I will address any further feedback in a follow-up PR along with the API review feedback.

Thank you @SteveSandersonMS and @javiercn for your detailed feedback and collaboration on this!
